File: Chapter06/Java/CommonControls/app/src/main/java/com/androidbook/commoncontrols/ListViewActivity3.java
package com.androidbook.commoncontrols;

import android.app.ListActivity;
import android.database.Cursor;
import android.os.Bundle;
import android.provider.Contacts.People;
import android.provider.ContactsContract;
import android.util.Log;
import android.util.SparseBooleanArray;
import android.view.View;
import android.widget.ListView;
import android.widget.SimpleCursorAdapter;

public class ListViewActivity3 extends ListActivity {
    private static final String TAG = "ListViewActivity3";
    private ListView lv = null;
    private Cursor cursor = null;
    private int idCol = -1;
    private int nameCol = -1;
    private int notesCol = -1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.list);

        lv = getListView();

        cursor = getContentResolver().query(ContactsContract.Contacts.CONTENT_URI,
                null, null, null, ContactsContract.Contacts.DISPLAY_NAME);

        String[] cols = new String[]{ContactsContract.Contacts.DISPLAY_NAME};
        idCol = cursor.getColumnIndex(ContactsContract.Contacts._ID);
        nameCol = cursor.getColumnIndex(ContactsContract.Contacts.DISPLAY_NAME);

        int[] views = new int[]{android.R.id.text1};

        SimpleCursorAdapter adapter = new SimpleCursorAdapter(this,
                android.R.layout.simple_list_item_multiple_choice,
                cursor, cols, views);

        this.setListAdapter(adapter);

        lv.setChoiceMode(ListView.CHOICE_MODE_MULTIPLE);
    }

    public void doClick(View view) {
        int count = lv.getCount();
        SparseBooleanArray viewItems = lv.getCheckedItemPositions();
        for (int i = 0; i < count; i++) {
            if (viewItems.get(i)) {
                // CursorWrapper cw = (CursorWrapper) lv.getItemAtPosition(i);
                cursor.moveToPosition(i);
                long id = cursor.getLong(idCol);
                String name = cursor.getString(nameCol);
                Log.v(TAG, name + " is checked. Position = " + i + ". Id = " + id);
            }
        }
    }
}

File: Chapter06/Kotlin/CommonControls/app/src/main/java/com/androidbook/commoncontrols/ListViewActivity3.kt
package com.androidbook.commoncontrols

import android.database.Cursor
import android.os.Bundle
import android.provider.ContactsContract
import android.util.Log
import android.view.View
import android.widget.ListView
import android.widget.SimpleCursorAdapter
import androidx.appcompat.app.AppCompatActivity

class ListViewActivity3: AppCompatActivity() {

    private lateinit var lv: ListView
    private var cursor: Cursor? = null

    private var idCol = -1
    private var nameCol = -1
    private var notesCol = -1

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.list)

        lv = findViewById(android.R.id.list)

        cursor = contentResolver.query(ContactsContract.Contacts.CONTENT_URI,
            null, null, null, ContactsContract.Contacts.DISPLAY_NAME)

        val cols = arrayOf(ContactsContract.Contacts.DISPLAY_NAME)
        cursor?.let {
            idCol = it.getColumnIndex(ContactsContract.Contacts._ID)
            nameCol = it.getColumnIndex(ContactsContract.Contacts.DISPLAY_NAME)
        }

        val views = intArrayOf(android.R.id.text1)

        val adapter = SimpleCursorAdapter(this,
            android.R.layout.simple_list_item_multiple_choice,
            cursor, cols, views
        )

        lv.adapter = adapter

        lv.choiceMode = ListView.CHOICE_MODE_MULTIPLE

    }

    public fun doClick(view: View) {
        val count = lv.count
        val viewItems = lv.checkedItemPositions
        for (i in 0..count) {
            if (viewItems[i]) {
//                val cw = lv.getItemAtPosition(i) as CursorWrapper
                cursor?.let {
                    it.moveToPosition(i)
                    val id = it.getLong(idCol)
                    val name = it.getString(nameCol)
                    Log.v(TAG, "$name is checked. Position = $i. Id = $id")
                }
            }
        }
    }

    companion object {
        const val TAG = "ListViewActivity3"
    }
}